### PR TITLE
Clarify requisition refusal justifications.

### PR DESCRIPTION
**Pull request by SanjayVas**
Some of the values did not make sense in the v2alpha API.

---
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/25)

### CODE REVIEW

**Review by zachcwc — APPROVED** [review]

**Review by kungfucraig — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/25)* status: 0 of 1 files reviewed, 6 unresolved discussions (waiting on @oliver-amzn and @SanjayVas)

___
*[src/main/proto/wfa/measurement/api/v2alpha/requisition.proto, line 116 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/25#-MZtZtDP7fari0ZvzMNs:-MZtZtDP7fari0ZvzMNt:b-gvd8dr) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/177454f19a4cc60c5a7452a31461664a139d2c5d/src/main/proto/wfa/measurement/api/v2alpha/requisition.proto#L116)):*
> ```protobuf
>   message Refusal {
>     // Possible justification for refusing to fulfill a `Requisition`.
>     enum Justification {
> ```

I find myself questioning the need for this at all. It's probably only useful if we want to keep stats for particular refusal reasons or do some kind of programatic recovery.

It's not clear to me that every value in the enum meets one of those criteria. 

The PB one makes sense. The signature one does too because of the complexity of our crypto. Many of the other ones could be rolled into "CLIENT_INPUT_ERROR" or something like that.

Presumably none of these are retryable. Maybe document this? Maybe it's obvious though...
___
*[src/main/proto/wfa/measurement/api/v2alpha/requisition.proto, line 122 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/25#-MZtZEjE78rSLMRBJ5sA:-MZtZEjE78rSLMRBJ5sB:bcjavcr) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/177454f19a4cc60c5a7452a31461664a139d2c5d/src/main/proto/wfa/measurement/api/v2alpha/requisition.proto#L122)):*
> ```protobuf
>       //
>       // This should only happen if the corresponding event group in the
>       // `DataProvider`'s system previously existed but has been deleted.
> ```

Couldn't this happen if the measurement consumer puts some random number in the event group field?
___
*[src/main/proto/wfa/measurement/api/v2alpha/requisition.proto, line 127 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/25#-MZtZOml-pZdnexImTMp:-MZtZOml-pZdnexImTMq:bdepyzn) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/177454f19a4cc60c5a7452a31461664a139d2c5d/src/main/proto/wfa/measurement/api/v2alpha/requisition.proto#L127)):*
> ```protobuf
>       // `measurement_spec`, `encrypted_requisition_spec`, or `protocol_config`
>       // specifies a configuration that the `DataProvider` does not support.
>       CONFIGURATION_UNSUPPORTED = 2;
> ```

Consider dropping "UNKNOWN_EVENT_GROUP" and rolling that case into this one.
___
*[src/main/proto/wfa/measurement/api/v2alpha/requisition.proto, line 131 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/25#-MZtZXhNAL7bZIKX_inm:-MZtZXhNAL7bZIKX_inn:b-55g5po) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/177454f19a4cc60c5a7452a31461664a139d2c5d/src/main/proto/wfa/measurement/api/v2alpha/requisition.proto#L131)):*
> ```protobuf
>       // The collection interval covers a range that is sufficiently long ago
>       // that the `DataProvider` no longer retains the underlying data.
>       COLLECTION_INTERVAL_TOO_DISTANT = 3;
> ```

What if the interval includes the future? Do you want to make this more general to include that too?

This could also be rolled into CONFIGURATION_UNSUPPORTED (aka CLIENT_ERROR or INPUT_ERROR).
___
*[src/main/proto/wfa/measurement/api/v2alpha/requisition.proto, line 147 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/25#-MZtaEYW9HNMDpJHrln5:-MZtaEYW9HNMDpJHrln6:b-6t1clb) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/177454f19a4cc60c5a7452a31461664a139d2c5d/src/main/proto/wfa/measurement/api/v2alpha/requisition.proto#L147)):*
> ```protobuf
>       // The `DataProvider` has declined to fulfill this `Requisition`
>       // regardless of any of the other `Justification` conditions.
>       DECLINED = 7;
> ```

I don't know how DECLINED and UNFULFILLABLE differ. They are both opaque and either there will be a message with more details or not.

They seem distinct from the other cases above though.
___
*[src/main/proto/wfa/measurement/api/v2alpha/requisition.proto, line 159 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/25#-MZtaW_Z8RNqRubnHcs4:-MZtaW_Z8RNqRubnHcs5:bgtgw5d) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/177454f19a4cc60c5a7452a31461664a139d2c5d/src/main/proto/wfa/measurement/api/v2alpha/requisition.proto#L159)):*
> ```protobuf
>   }
>   // The refusal that put this `Requisition` into the `REFUSED` state. Must be
>   // set when `state` is `REFUSED`.
> ```

Can you add "Repeated attempts to fulfill a refused requisition will also be refused." or "A refusal indicates a permanent error condition." or similar.

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/25)* status: 0 of 1 files reviewed, 6 unresolved discussions (waiting on @kungfucraig and @oliver-amzn)

___
*[src/main/proto/wfa/measurement/api/v2alpha/requisition.proto, line 116 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/25#-MZtZtDP7fari0ZvzMNs:-MZu44S49ZguFG2iI8a1:b-8a34io) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/177454f19a4cc60c5a7452a31461664a139d2c5d/src/main/proto/wfa/measurement/api/v2alpha/requisition.proto#L116)):*
<details><summary><i>Previously, kungfucraig (Craig Wright) wrote…</i></summary><blockquote>

I find myself questioning the need for this at all. It's probably only useful if we want to keep stats for particular refusal reasons or do some kind of programatic recovery.

It's not clear to me that every value in the enum meets one of those criteria. 

The PB one makes sense. The signature one does too because of the complexity of our crypto. Many of the other ones could be rolled into "CLIENT_INPUT_ERROR" or something like that.

Presumably none of these are retryable. Maybe document this? Maybe it's obvious though...
</blockquote></details>

Collapsed these per offline discussion and added some explanation of the organization.
___
*[src/main/proto/wfa/measurement/api/v2alpha/requisition.proto, line 122 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/25#-MZtZEjE78rSLMRBJ5sA:-MZu4Lby6dXYKiLDrHpP:b-t1f39r) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/177454f19a4cc60c5a7452a31461664a139d2c5d/src/main/proto/wfa/measurement/api/v2alpha/requisition.proto#L122)):*
<details><summary><i>Previously, kungfucraig (Craig Wright) wrote…</i></summary><blockquote>

Couldn't this happen if the measurement consumer puts some random number in the event group field?
</blockquote></details>

This was assuming that the EventGroup Key had already been checked against the known/registered ones. The Kingdom cannot enforce this in the v2alpha design. Removed.
___
*[src/main/proto/wfa/measurement/api/v2alpha/requisition.proto, line 127 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/25#-MZtZOml-pZdnexImTMp:-MZu4oTk7W8AQUrRnXhO:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/177454f19a4cc60c5a7452a31461664a139d2c5d/src/main/proto/wfa/measurement/api/v2alpha/requisition.proto#L127)):*
<details><summary><i>Previously, kungfucraig (Craig Wright) wrote…</i></summary><blockquote>

Consider dropping "UNKNOWN_EVENT_GROUP" and rolling that case into this one.
</blockquote></details>

Done.
___
*[src/main/proto/wfa/measurement/api/v2alpha/requisition.proto, line 131 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/25#-MZtZXhNAL7bZIKX_inm:-MZu4pMc3Lp6YPfTVi6h:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/177454f19a4cc60c5a7452a31461664a139d2c5d/src/main/proto/wfa/measurement/api/v2alpha/requisition.proto#L131)):*
<details><summary><i>Previously, kungfucraig (Craig Wright) wrote…</i></summary><blockquote>

What if the interval includes the future? Do you want to make this more general to include that too?

This could also be rolled into CONFIGURATION_UNSUPPORTED (aka CLIENT_ERROR or INPUT_ERROR).
</blockquote></details>

Done.
___
*[src/main/proto/wfa/measurement/api/v2alpha/requisition.proto, line 147 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/25#-MZtaEYW9HNMDpJHrln5:-MZu4qXU1IYUUFQhMcN6:b-oo5bpr) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/177454f19a4cc60c5a7452a31461664a139d2c5d/src/main/proto/wfa/measurement/api/v2alpha/requisition.proto#L147)):*
<details><summary><i>Previously, kungfucraig (Craig Wright) wrote…</i></summary><blockquote>

I don't know how DECLINED and UNFULFILLABLE differ. They are both opaque and either there will be a message with more details or not.

They seem distinct from the other cases above though.
</blockquote></details>

`UNFULFILLABLE` is for when the `DataProvider` *should* be able to fulfill it, but cannot for some reason.

`DECLINED` is for when the `DataProvider` might be able to fulfill it, but doesn't want to (e.g. they have a policy of only fulfilling above some size threshold).

Added some more comments to clarify.
___
*[src/main/proto/wfa/measurement/api/v2alpha/requisition.proto, line 159 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/25#-MZtaW_Z8RNqRubnHcs4:-MZu5_ta-9rU7SMYANkx:btccljh) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/177454f19a4cc60c5a7452a31461664a139d2c5d/src/main/proto/wfa/measurement/api/v2alpha/requisition.proto#L159)):*
<details><summary><i>Previously, kungfucraig (Craig Wright) wrote…</i></summary><blockquote>

Can you add "Repeated attempts to fulfill a refused requisition will also be refused." or "A refusal indicates a permanent error condition." or similar.
</blockquote></details>

`REFUSED` is already listed as a terminal state for a `Requisition`.

Separately, it's not necessarily true that creating a new `Measurement` (and therefore new `Requisition`) with the same specification would be refused again, though that would be the case most of the time. For example, if an EDP manages to recover data that was thought lost, changes a policy, or just fixes a bug.

**Review by oliver-amzn — APPROVED** [review]
> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/25)* status: all files reviewed, 6 unresolved discussions (waiting on @kungfucraig)

File: src/main/proto/wfa/measurement/api/v2alpha/requisition.proto [hunk 1]
```diff
@@ -112,39 +112,50 @@ message Requisition {
   bytes data_provider_participation_signature = 11;
 
   message Refusal {
+    // Justification for refusing to fulfill a `Requisition`.
+    //
+    // The number of values should be kept small and communicate only the
+    // granularity necessary for automated downstream consumption. Each value
```
Comment (kungfucraig) [inline src/main/proto/wfa/measurement/api/v2alpha/requisition.proto:118]: or stats gathering

**Review by kungfucraig — APPROVED** [review]

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/25)* status: 0 of 1 files reviewed, 7 unresolved discussions (waiting on @kungfucraig and @oliver-amzn)

___
*[src/main/proto/wfa/measurement/api/v2alpha/requisition.proto, line 118 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/25#-MZsoG6n-XRpFylVykk1-r2-118:-MZxNux26tEqo6_ilBSC:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/54fbf9fb65437d91bd6b42980a1ab6a131097c4f/src/main/proto/wfa/measurement/api/v2alpha/requisition.proto#L118)):*
<details><summary><i>Previously, kungfucraig (Craig Wright) wrote…</i></summary><blockquote>

or stats gathering
</blockquote></details>

Done.

**Pull request merged**